File: starters/cs/Map.cs
using System;

class Map
{
  // Stores the width and height of the Tron map.
  private static int width, height; 
  
  // Stores the actual contents of the Tron map.
  private static bool[,] walls;
  
  // Stores the locations of the two players.
  private static Point myLocation, opponentLocation;

  public static int Width()
  {
   return width;
  }
  
  public static int Height()
  {
   return height;
  }
  
  public static bool IsWall(int x, int y) 
  {
	if (x < 0 || y < 0 || x >= width || y >= height) 
	{
	    return true;
	} 
	else 
	{
	    return walls[ x, y ];
	}
  }
  
  // My X location.
  public static int MyX() 
  {
	    return (int)myLocation.X;
  }

  // My Y location.
  public static int MyY() 
  {
	    return (int)myLocation.Y;
  }  
  
  // The opponent's X location.
  public static int OpponentX() 
  {
	    return (int)opponentLocation.X;
  }

  // The opponent's Y location.
  public static int OpponentY() 
  {
	    return (int)opponentLocation.Y;
  }
  
  // Transmit a move to the game engine. 'direction' can be any sort of
  // string that indicates a direction. For example, "north", "North",
  // "n", "N".
  public static void MakeMove(string direction) 
  {
   if (direction.Length <= 0) {
	    Console.Error.WriteLine("FATAL ERROR: empty direction string. You " +
			       "must specify a valid direction in which " +
			       "to move.");
	    Environment.Exit(1);
   }
   string temp = direction.Substring(0, 1).ToUpper(); 
   int firstChar = (int)temp[0];
   
	switch (firstChar) {
	case 'N':
	    MakeMove(1);
	    break;
	case 'E':
	    MakeMove(2);
	    break;
	case 'S':
	    MakeMove(3);
	    break;
	case 'W':
	    MakeMove(4);
	    break;
	default:
	    Console.Error.WriteLine("FATAL ERROR: invalid move string. The string must " +
			       "begin with one of the characters 'N', 'E', 'S', or " +
			       "'W' (not case sensitive).");
	    Environment.Exit(1);
	    break;
	}
  }
  
     // Reads the map from standard input (from the console).
  public static void Initialize() 
  {
   string firstLine = "";
   try {
	     int c;
	     while ((c = Console.Read()) >= 0) 
	     {
		  if (c == '\n') {
		    break;
		  }
		  firstLine += (char)c;
	    }
	} 
	catch (Exception e) 
	{
	    Console.Error.WriteLine("Could not read from stdin.");
	    Environment.Exit(1);
	}
	firstLine = firstLine.Trim();
	if (firstLine.Equals("") || firstLine.Equals("exit")) 
	{
	    Environment.Exit(1); // If we get EOF or "exit" instead of numbers
	                    // on the first line, just exit. Game is over.
	}
	string[] tokens = firstLine.Split(' ');
	if (tokens.Length != 2) {
	    Console.Error.WriteLine("FATAL ERROR: the first line of input should " +
			       "be two integers separated by a space. " +
			       "Instead, got: " + firstLine);
	    Environment.Exit(1);
	}
	try {
	    width = Convert.ToInt32(tokens[0]);
	    height = Convert.ToInt32(tokens[1]);
	} catch (Exception e) {
	    Console.Error.WriteLine("FATAL ERROR: invalid map dimensions: " +
			       firstLine);
	    Environment.Exit(1);
	}
	walls = new bool[width,height];
	bool foundMyLocation = false;
	bool foundHisLocation = false;
	int numSpacesRead = 0;
	int x = 0, y = 0;
	while (y < height) {
	    int c = 0;
	    try {
		 c = Console.Read();
	} catch (Exception e) {
		Console.Error.WriteLine("FATAL ERROR: exception while reading " +
				   "from stdin.");
		Environment.Exit(1);
	}
	if (c < 0) {
		break;
	}
	switch (c) 
	{
	case '\n':
	if (x != width) {
	   Console.Error.WriteLine("Invalid line length: " + x + "(line " + y + ")");
		Environment.Exit(1);
   }
	++y;
	x = 0;
	continue;
	case '\r':
	continue;
	case ' ':
	walls[x,y] = false;
	break;
	case '#':
	walls[x,y] = true;
	break;
	case '1':
	if (foundMyLocation) {
	   Console.Error.WriteLine("FATAL ERROR: found two locations " +
			                     "for player " +
				                  "1 in the map! First location is (" +
				                  myLocation.X + "," +
				                  myLocation.Y +
				                  "), second location is (" + x + "," +
				                  y + ").");
		Environment.Exit(1);
   }
	walls[x,y] = true;
	myLocation = new Point(x, y);
	foundMyLocation = true;
	break;
	case '2':
	if (foundHisLocation) {
	    Console.Error.WriteLine("FATAL ERROR: found two locations for player " +
				       "2 in the map! First location is (" +
				       opponentLocation.X + "," +
				       opponentLocation.Y + "), second location " +
				       "is (" + x + "," + y + ").");
		 Environment.Exit(1);
	}
	walls[x,y] = true;
	opponentLocation = new Point(x, y);
	foundHisLocation = true;
	break;
	default:
	Console.Error.WriteLine("FATAL ERROR: invalid character received. " +
				   "ASCII value = " + c);
   Environment.Exit(1);
   break;
	}
	 ++x;
	 ++numSpacesRead;
	}
	if (numSpacesRead != width * height) {
	    Console.Error.WriteLine("FATAL ERROR: wrong number of spaces in the map. " +
			       "Should be " + (width * height) + ", but only found " +
			       numSpacesRead + " spaces before end of stream.");
	    Environment.Exit(1);
	}
	if (!foundMyLocation) {
	    Console.Error.WriteLine("FATAL ERROR: did not find a location for player 1!");
	    Environment.Exit(1);
	}
	if (!foundHisLocation) {
	    Console.Error.WriteLine("FATAL ERROR: did not find a location for player 2!");
	    Environment.Exit(1);
	}
 }
// Writes the given integer (direction code) to stdout.
    //   1 -- North
    //   2 -- East
    //   3 -- South
    //   4 -- West
 private static void MakeMove(int direction) {
	       Console.WriteLine(direction);
 }
}

class Point{
public int X, Y;
public Point(int x, int y)
{
 X = x;
 Y = y;
}
}

File: starters/ruby/map.rb
# The Map class gets the Tron map from the game engine, and provides some
# methods that allow you to look at it.
#
# You should not change any of the code in this file. It's just here for
# your convenience, to do the boring parts for you.
#
# Copied from the Java example

class Map

	attr_reader :width, :height, :my_position, :opponent_position
	
	def initialize()
	
		@width = -1
		@height = -1
		@walls = []
		@my_position = [-1,-1]
		@opponent_position = [-1,-1]
		
		read_map
		
	end	
	
	def read_map
	
		begin
		
			#read the width and height from the first line
			firstline = $stdin.readline("\n")
			width, height = firstline.split(" ")
			@width = width.to_i
			@height = height.to_i
			
			#check for properly formatted width, height
			if height == 0 or width == 0
				p "OOPS!: invalid map dimensions: " + firstline
				exit(1)
			end
			
			#read the representation of the board
			lines = []
			@height.times do
				lines += [$stdin.readline("\n").strip]
			end
			board = lines.join("")
			
			#get the wall data
			@walls = board.split(//).map{|char| char == " " ? false : true}
			
			#get player starting locations
			p1start = board.index("1").to_i
			p2start = board.index("2").to_i
			
			if board.split(//).select{|char| char == "1"}.size > 1
				p "OOPS!: found more than 1 location for player 1"
				exit(1)
			end
			
			if board.split(//).select{|char| char == "2"}.size > 1
				p "OOPS!: found more than 1 location for player 2"
				exit(1)
			end
			
			p "OOPS!: Cannot find locations." if p1start == nil or p2start == nil
			
			pstartx = p1start % @width
			pstarty = (p1start / @width)
			@my_position = [pstartx, pstarty]
			
			pstartx = p2start % @width
			pstarty = (p2start / @width)
			@opponent_position = [pstartx, pstarty]
					
			
		rescue EOFError => e
			# Got EOF: tournament is finished.
			exit(0)
			
		rescue => e
			p  e
			exit(1)
		end
	
	end
	private :read_map
	
	def each(&proc)
		
		(0..@height-1).each{|y|
			(0..@width-1).each{|x|
				proc[x, y, wall?(x, y)]
			}
		}
		
	end
	
	
	def wall? (x, y)
		return true if x < 0 or y < 0 or x >= @width or y >= @height
		return @walls[x+@width*y]
	end
	
	def to_string()

		out = ""
		counter = 0
				
		@height.times do
			@width.times do
				out += @walls[counter] == true ? "#" : "-"
				counter+=1
			end
			out += "\n"
		end
		
		
		return out
		
	end



	def make_move(direction)
	
		$stdout << ({:NORTH=>1, :SOUTH=>3, :EAST=>2, :WEST=>4}[direction])
		$stdout << "\n"
		$stdout.flush
		
	end
	
end
